### DISCUSSION TITLE
Removing text from every page of a PDF

### DISCUSSION THREAD

**Discussion by j-t-1** [py-pdf/pypdf#2489] [Q&A] [answered]
I am trying to remove two lines of text at the bottom of every page of a PDF.
```
from pypdf import PdfReader, PdfWriter

with open("example.pdf" "rb") as f0:
    reader = PdfReader(f0)

    for page in reader.pages:
        page.cropbox.lower_left = (0, 25)
        page.cropbox.lower_right = (595.276, 25)

    writer = PdfWriter(clone_from=reader)

    with open("out.pdf", "wb") as f1:
        writer.write(f1)
```
1. The above code is not working; is there an obvious fault or a better way?
2. Can the text be removed directly (something like deflate the PDF, then view
it to see which object contains the text, then deleting the references)?

**Comment by stefan6419846** [comment, marked as answer]
> The above code is not working; is there an obvious fault or a better way?

What do you mean by "not working"? Please note that just moving the box does not guarantee that the text will disappear, but maybe just hide it.

> Can the text be removed directly (something like deflate the PDF, then view it to see which object contains the text, then deleting the references)?

This highly depends on the PDF itself, although it generally should be doable if the PDF file has a valid text layer. You probably want to have a look at `page.get_contents` and `page.replace_contents` for this.

**Comment by j-t-1** [comment]
Apologies for the ridiculously vague "not working"; I was expecting 25/72 inch to be clipped (not visible) from the bottom of each page.

This variation tries (but fails) to remove an inch from the bottom of the page:
```python
from pypdf import PdfReader, PdfWriter

with open("input.pdf", "rb") as f0:
    reader = PdfReader(f0)

    for page in reader.pages:
        page.cropbox.lower_left = (0, 72)
        page.cropbox.upper_right = (595.276, 842)

    writer = PdfWriter(clone_from=reader)

    with open("out.pdf", "wb") as f1:
        writer.write(f1)
```
When using ```page.get_contents```, is there a technique to find the resource containing the text wanted for removal?

**Reply by stefan6419846** [reply]
When using `page.get_contents()` (or `page._get_contents_as_bytes()`), you need to be able to understand the PDF internals regarding the text operators (`BT`, `ET`, `TJ`, `Tj` and so on) - how easy this is, hardly depends on the specific file.

In my case (with an arbitrary landscape mode PDF file of one page) using

```python3
writer = PdfWriter(clone_from="input.pdf")

for page in writer.pages:
    print(page.mediabox, page.trimbox, page.artbox, page.cropbox)
    page.cropbox.lower_left = (0, 72)
    print(page.mediabox, page.trimbox, page.artbox, page.cropbox)

writer.write("output.pdf")
```

correctly removes the lower part of the page.

**Reply by j-t-1** [reply]
Your code setting the cropbox works nicely.
Looping through the PageObject and calling member ```page.get_contents()``` does return a ```pypdf.generic._data_structures.ContentStream``` but it is always empty, which is allowed since /Contents is optional. Is /Contents metadata-like, used for something like indexing? Is there another function to do a replacement, maybe method_replace_object of PdfWriter?

**Reply by stefan6419846** [reply]
Please note that I do not recommend to rely on internal APIs like `_replace_object` or `_get_contents_as_bytes` - although they might appear stable, there is no guarantee that they will remain like this and not break at any time. Relying on public APIs is preferred for these reasons.

`/Contents` is documented inside the PDF specification as "stream or array" with the following value:

> (Optional) A content stream (see 7.8.2, "Content Streams") that shall describe the contents of this page. If this entry is absent, the page shall be empty.
>
> The value shall be either a single stream or an array of streams. If the value is an array, the effect shall be as if all of the streams in the array were concatenated, in order, to form a single stream. Conforming writers can create image objects and other resources as they occur, even though they interrupt the content stream. The division between streams may occur only at the boundaries between lexical tokens (see 7.2, "Lexical Conventions") but shall be unrelated to the page’s logical content or organization. Applications that consume or produce PDF files need not preserve the existing structure of the Contents array. Conforming writers shall not create a Contents array containing no elements.

**Reply by pubpub-zz** [reply]
> Your code setting the cropbox works nicely.
> Looping through the PageObject and calling member ```page.get_contents()``` does return a ```pypdf.generic._data_structures.ContentStream``` but it is always empty, which is allowed since /Contents is optional. Is /Contents metadata-like, used for something like indexing? Is there another function to do a replacement, maybe method_replace_object of PdfWriter?

Be aware that direct display of contents will most probably report an 'empty' dict but to see actual content call `contentobj.get_data()`

**Reply by j-t-1** [reply]
The call to ```get_data()``` worked, and when printed includes:
/Artifact BMC\n\nq 1 0 0 1 10 0 cm\nq 0.52941 0.52941 0.52941 rg BT/F1 8 Tf 10.16 TL 0 12.416 Td 0 Tc 178.162 0 Td(Text to remove)
Is there an example of how to use this in ```replace_contents()```?

**Reply by stefan6419846** [reply]
Like `content_stream.get_data()`, there also is `content_stream.set_data(data)` to update the data. As `page.get_contents()` generates the `ContentStream` on the fly, you will have to call `page.replace_contents(content_stream)` to modify the actual page as well.

**Reply by j-t-1** [reply]
```
writer = PdfWriter(clone_from="input.pdf")
with open("output.pdf", "wb") as output:
    for page in writer.pages:
        content_stream = page.get_contents()
        content_stream.set_data(content_stream.get_data()[:-281])
        page.replace_contents(content_stream)
    writer.write(output)
```
Works, thank you for your help.

I also tried the ```clone``` method of ```ContentStream``` using the ```ignore_fields``` parameter but did not get it working. I have a few PDFs wanting the same text removed, and do not know if the text is always the last ~281 characters, and so the ```clone``` method may be the best solution.

**Reply by stefan6419846** [reply]
Apart from a *pypdf* solution, you should be able cover these edge cases with plain Python functionality by using the generic `bytes.find()` or `bytes.index()` methods. This will report the start position of your data and allows you to slice the data accordingly. With this approach you can introduce dedicated handling for all cases where the desired source snippet is missing as well.

**Reply by pubpub-zz** [reply]
clone function is designed to clone an object into a pdfWriter. If you clone multiple times the same object, you will get always the same object(this is designed to match call to sub objects). If you want a real new you should set `force_duplicate=True`

**Reply by j-t-1** [reply]
The non-pypdf solution of find and replace is a good idea I had not thought of, I tried searching for the string but it seems fragmented.

Trying the clone function of ContentStream is below:
```
writer = PdfWriter(clone_from="input.pdf")
with open("output.pdf", "wb") as output:
    for page in writer.pages:
        content_stream = page.get_contents()
        content_stream = content_stream.clone(output, force_duplicate = True, ignore_fields = ("/Artifact"))
        page.replace_contents(content_stream)
    writer.write(output)
```

This caused a RecursionError: maximum recursion depth exceeded while calling a Python object. The most recent call was in _utils.py", line 235, in read_until_regex stream.seek(m.start() - (len(name) + len(tok)), 1).

The PDF is over 1000 pages, so likely a factor.

1. How do I know what to put as the ignore_fields. The text to remove is something like:
/Artifact BMC\n\nq 1 0 0 1 10 0 cm\nq 0.52941 0.52941 0.52941 rg BT/F1 8 Tf 10.16 TL 0 12.416 Td 0 Tc 178.162 0 Td(Text to remove)

2. In the documentation of pypdf.generic.ContentStream what does this mean:
In order to be fast, this datastructure can contain either: * raw data in ._data * parsed stream operations in ._operations?

**Reply by pubpub-zz** [reply]
@j-t-1 
can you provide the document or an extract of a few pages please

**Reply by j-t-1** [reply]
@pubpub-zz alas no, something similar are the download as PDF Wikibooks. The Wikibooks last page has in gray something like the following (very similar to the PDFs I am working with, but at the bottom of every page):

> This page was last edited on 29 November 2023, at 00:22.
> Text is available under the Creative Commons Attribution-ShareAlike License; additional terms may apply. By using
> this site, you agree to the Terms of Use and Privacy Policy.
>

**Reply by stefan6419846** [reply]
> In the documentation of pypdf.generic.ContentStream what does this mean:
> 
> > In order to be fast, this datastructure can contain either: * raw data in ._data * parsed stream operations in ._operations?

Raw data is a byte string with the operations, the parsed variant is a list of single operations represented by their byte strings. This fact should be mostly irrelevant for you as long as you use the public functionality from `ContentStream.get_data()`.

**Reply by j-t-1** [reply]
After commenting out all the other lines, this causes the error:

```
writer = PdfWriter(clone_from="input.pdf")
```

```
Traceback (most recent call last):
  File "/home/work/.local/lib/python3.10/site-packages/pypdf/generic/_data_structures.py", line 423, in read_from_stream
    value = read_object(stream, pdf, forced_encoding)
  File "/home/work/.local/lib/python3.10/site-packages/pypdf/generic/_data_structures.py", line 1225, in read_object
    return NumberObject.read_from_stream(stream)
  File "/home/work/.local/lib/python3.10/site-packages/pypdf/generic/_base.py", line 459, in read_from_stream
    return NumberObject(num)
  File "/home/work/.local/lib/python3.10/site-packages/pypdf/generic/_base.py", line 425, in __new__
    return int.__new__(cls, int(value))
RecursionError: maximum recursion depth exceeded while calling a Python object
```

Small files work fine, but not a 1000+ page ~1.1M PDF.

**Reply by pubpub-zz** [reply]
I'm not sure this is due to size. Can you please provide the pdf for analysis

**Reply by j-t-1** [reply]
At the moment no. I will test with other large files and develop a hypothesis.

**Reply by j-t-1** [reply]
This gives the same error as `writer = PdfWriter(clone_from="input.pdf")`:

```
reader = PdfReader("input.pdf")
writer = PdfWriter()

# Does not work, but does if 4 is reduced to 3
for page in reader.pages[:4]:
    writer.add_page(page)`
```
@pubpub-zz I tried different PDFs with more pages, and they worked.

**Reply by stefan6419846** [reply]
Then this tends to be related to the corresponding page and requires further analysis on the actual file.

**Reply by j-t-1** [reply]
Finally got this working.
```
setrecursionlimit(10000)

writer = PdfWriter(clone_from='input.pdf')

for page in writer.pages:
    contents = page._get_contents_as_bytes()
    contents = contents[:contents.find(b'/Artifact BMC\n') - 1]
    stream = StreamObject()
    stream.set_data(contents)
    page.replace_contents(stream)

with open("output.pdf", "wb") as f:
    writer.write(f)
```
Thanks @pubpub-zz and @stefan6419846.
